docs/design/features/lsra-detail.md

```diff
@@ -907,19 +924,162 @@ Future Extensions and Enhancements
 
-The following haven't yet been opened as github issues:
+## Code Quality Enhancements
+
+### <a name="combine"/>Merge Allocation of Free and Busy Registers
+
+This is captured as [\#15408](https://github.com/dotnet/coreclr/issues/15408)
+Consider merging allocating free & busy regs.
+
+Currently the register allocator will always allocate an available register, even if it only meets
+the minimum requirements for the current node. This results in some really bad worst-case behavior
+where a single scratch register is available, and is immediately spilled because it is killed prior
+to the next use. Often this same register will be used and spilled many times before a better register
+becomes free.
+
+The alternative approach under consideration is to combine free and busy register allocation
+(`tryAllocateFreeReg()` and `allocateBusyReg()`) such that a busy register will be spilled if there
+are no suitable free registers, and the current `Interval` has greater weight than the `Interval`
+occupying the register. This must be accompanied by some improvements in the efficiency of the
+checks, so as not to degrade throughput. This is currently a work-in-progress (https://github.com/CarolEidt/coreclr/tree/CombineAlloc) but hasn't yet been ported to the runtime repo, and needs
+further work to eliminate diffs and improve throughput.
+
+This would make it possible to spill a register for a higher weight `lclVar` rather than "settling"
+for a register that's a poor fit for its requirements. This is probably the best approach to
+address Issues [\#7664](https://github.com/dotnet/coreclr/issues/7664)
+Heuristics for callee saved reg allocation and
+[\#13735](https://github.com/dotnet/coreclr/issues/13735)
+Let variables within a loop use register first.
+
+The following issues are related:
+
+* Both `tryAllocateFreeReg()` and `allocateBusyReg()` currently fully evaluate the "goodness" of each register.
+  Issue [\#7301](https://github.com/dotnet/coreclr/issues/7301) tracks the possibility of short-circuiting
+  this evaluation.
+  Making such an improvement should probably be done in conjunction with this work.
+
+* Issue [\#26847](https://github.com/dotnet/coreclr/issues/26847)
+  Heuristics for callee saved reg allocation.
+
+### Auto-tuning of register selection
+
+This is not yet captured as a GitHub issue.
+
+This would be best done after [Merge Allocation of Free and Busy Registers](#combine).
+
+The idea would be to add support to change the weight given to the various selection heuristics
+according to a configuration specification, allowing them to be auto-tuned.
+
+The scoring could be done using perf scores, or even using actual performance measurements (though the
+former would be considerably more efficient).
```

Comment:
> though the former would be considerably more efficient

Why do you think that static counters will give a better picture than the actual numbers collected during Tier-0?

Reply:
I don't think they would give a better picture, but the process itself will be more efficient - that is, it would be much more costly to do an auto-tuning where each configuration had to be scored using an actual benchmark run. I'll try to clarify.

Reply:
The tradeoffs here are interesting:
* Use the current PerfScore that is noise-free and can be easily captured for all methods,  but is possibly not well-correlated with perf, or
* Use some kind of sampled time estimate that is fairly noisy, only can be captured for methods that get executed via test runs, and for most of those the measurements won't rise above the noise floor, leaving you with a small set of good training instances, or
* Do some kind of instrumented block counting that is in principle noise-free (provided your test cases are deterministic) and correlates well with the IR you're impacting (so instrumentation would possibly need to be done late, just before LSRA, unlike IBC) and use that to produce a refined dynamic PerfScore like metric. This would only be available for methods that you can execute, so would be sparse, but given that it's noise free you'd have it for more methods than a time measurement.

I think the instrumented block count approach might be the most promising, though you could probably get the process working first using the current PerfScore, and see if you can automatically tune heuristics to improve PerfScores. If so, then you could see if that leads to measurably improved perf, and if it doesn't, you could consider the more refined version.

Reply:
I would say that the PerfScore approach can be readily adapted to whatever measured block counts are available, and it seems reasonable and desirable to integrate such measured block counts into the PerfScore machinery.